pipx/pipxrc.py

```diff
@@ -0,0 +1,196 @@
+import json
+from pathlib import Path
+from typing import List, Dict, NamedTuple, Any, Optional, TypeVar
+
+from pipx.Venv import PipxVenvMetadata
+
+
+PIPX_INFO_FILENAME = "pipxrc.json"
+
+
+class JsonEncoderPipx(json.JSONEncoder):
+    def default(self, obj):
+        # only handles what json.JSONEncoder doesn't understand by default
+        if isinstance(obj, Path):
+            return {"__type__": "Path", "__Path__": str(obj)}
+        return super().default(obj)
+
+
+def _json_decoder_object_hook(json_dict):
+    if json_dict.get("__type__", None) == "Path" and "__Path__" in json_dict:
+        return Path(json_dict["__Path__"])
+    return json_dict
+
+
+# Used for consistent types of multiple kinds
+Multi = TypeVar("Multi")
+
+
+class InjectedPackage(NamedTuple):
+    pip_args: List[str]
+    verbose: bool
+    include_apps: bool
+    include_dependencies: bool
+    force: bool
+
+
+class InstallOptions(NamedTuple):
+    pip_args: Optional[List[str]]
+    venv_args: Optional[List[str]]
+    include_dependencies: Optional[bool]
+
+
+class PipxrcInfo:
+    def __init__(self):
+        self.package_or_url: Optional[str] = None
+        self.install: InstallOptions = InstallOptions(
+            pip_args=None, venv_args=None, include_dependencies=None
+        )
+        self.venv_metadata: Optional[PipxVenvMetadata] = None
+        self.injected_packages: Optional[Dict[str, InjectedPackage]] = None
+        self._pipxrc_version: str = "0.1"
+
+    def to_dict(self) -> Dict[str, Any]:
+        venv_metadata: Optional[Dict[str, Any]]
+        injected_packages: Optional[Dict[str, Dict[str, Any]]]
+
+        if self.venv_metadata is not None:
+            venv_metadata = self.venv_metadata._asdict()
+        else:
+            venv_metadata = None
+        if self.injected_packages is not None:
+            injected_packages = {
+                k: v._asdict() for (k, v) in self.injected_packages.items()
+            }
+        else:
+            injected_packages = None
+
+        return {
+            "package_or_url": self.package_or_url,
+            "install": self.install._asdict(),
+            "venv_metadata": venv_metadata,
+            "injected_packages": injected_packages,
+            "pipxrc_version": self._pipxrc_version,
+        }
+
+    def from_dict(self, pipxrc_info_dict) -> None:
+        self.package_or_url = pipxrc_info_dict["package_or_url"]
+        self.install = InstallOptions(**pipxrc_info_dict["install"])
+        self.venv_metadata = PipxVenvMetadata(**pipxrc_info_dict["venv_metadata"])
+        self.injected_packages = {
+            k: InjectedPackage(**v)
+            for (k, v) in pipxrc_info_dict["injected_packages"].items()
+        }
+
+
+class Pipxrc:
+    def __init__(self, venv_dir: Path, read: bool = True):
+        self.venv_dir = venv_dir
+        self.pipxrc_info = PipxrcInfo()
+        if read:
+            self.read()
```

Comment:
There is always exactly one pipxrc.json file for each venv. I think it might be easier to manage the pipxrc information if it's automatically included and initialized in the `Venv` class. 

That way, instead of

```
    venv = Venv(venv_dir, verbose=verbose)
    venv.install_package(package, pip_args)

    if include_apps:
        _run_post_install_actions(
            venv,
            package,
            constants.LOCAL_BIN_DIR,
            venv_dir,
            include_dependencies,
            force=force,
        )
    pipxrc = Pipxrc(venv_dir)
    pipxrc.add_injected_package(
        package, pip_args, verbose, include_apps, include_dependencies, force
    )
    pipxrc.write()
```

you could do

```
    venv = Venv(venv_dir, verbose=verbose)
    venv.install_package(package, pip_args)
    venv.rc.add_injected_package(
        package, pip_args, verbose, include_apps, include_dependencies, force
    )

    if include_apps:
        _run_post_install_actions( 
            venv,
            package,
            constants.LOCAL_BIN_DIR,
            venv_dir,
            include_dependencies,
            force=force,
        )
     venv.rc.write()
```

Longer term I think it might be even cleaner to add new methods to the `Venv` class that would transparently update the `rc` file. Something like
```
venv.inject_package(...)  # this would automatically update inject the package AND update the rc file
```

Reply:
> The distinction between `Pipxrc` and `PipxrcInfo` is a little blurry to me. I would lean toward combining them into a single class if possible.

My thought was that PipxrcInfo was like a NamedTuple on steroids, mainly just a data structure with some interesting defaults, and custom serialization and deserialization code to handle all of the member NamedTuples inside.  I wanted it to be separate so that it alone would determine the format of `pipxrc.json`, and that code would be all in one place.

Whereas Pipxrc was the class that dealt with reading and writing cached info, and didn't necessarily need to know the format of the underlying file.  Indeed it could be an interface to the rest of the code that is `pipxrc.json` file-format-agnostic.

Reply:
On the topic of closer integration with `Venv.py`.  I agree it seems like the way to go.  I originally started doing something similar to this, but ran into problems with circular imports (pipxrc.py imports Venv.py and vice-versa.)

Reply:
Or do you mean just including all the code now in `pipxrc.py` in `Venv.py`?  That might work.

Reply:
I don't have strong feelings on it. I think it's fine as is.

Reply:
I see this was renamed as `PackageInfo`. I think this is more readable, thanks.